### DISCUSSION TITLE
PorousFlow - Capillary pressure formulation for carbon fibre resin injection.

### DISCUSSION THREAD

**Discussion by s-dunnim** [idaholab/moose#21000] [Q&A Modules: General] [answered]
I am using the porous flow module to model resin injection, at around 1 bar, into a dry carbon fibre skeleton, which is under vacuum. I am using a single phase unsaturated model for this, as there is only a negligible amount of air.
(P = porepressure, Pc = capillary pressure, Pi = injection pressure, S = saturation)

I am having issues with the capillary pressure formulation; in order to achieve S = 0 with any of the available Pc relationships, I have to use an initial condition P <= -Pc_max (as Pc = -P). Having negative P is clearly not realistic in a system where P is always >= 0 (P = 0 in the vacuum, 0 < P <= Pi in the resin).

- Is using capillary pressure to calculate saturation the right approach for this scenario? Most of the domain except the flow front should either be at 0 or 1 saturation; in both cases the notion of capillary pressure seems meaningless.
- If so, how can I overcome the issues described with the Pc and S relationship? I have considered subtracting Pc_max from P everywhere, as it is only really the pressure gradient driving the flow (so, for example, in the vacuum P=0-Pc_max, and at the injection site P=Pi-Pc_max). I have also considered writing a custom capillary pressure user object for calculating S, however I don't really have any data for how P(/Pc) in the flow front relates S (or indeed if there even is a valid relationship).
- If not, do you have any suggestions for a better approach? I am considering writing some simple custom kernels, materials etc to implement Darcy flow, and advect an S variable (following a Volume Of Fluid - like approach) instead of calculating it from capillary pressure.

See my [earlier question](https://github.com/idaholab/moose/discussions/20862) for more discussion on using the porous flow module with 0 saturation.

TIA

<details>
<summary>Example input file</summary>
<br>

```
# 2D single phase unsaturated porous flow - rectilinear validation case
# All units in SI unless otherwise stated

[Mesh]
    [plate]
        type = GeneratedMeshGenerator
        dim = 2
        xmin = 0
        xmax = 0.3
        nx = 30
        ymin = 0
        ymax = 0.1
        ny = 10
    []
[]

[UserObjects]
    [dictator]
        type = PorousFlowDictator
        porous_flow_vars = porepressure
        number_fluid_phases = 1
        number_fluid_components = 1
    []
    [pc]
        type = PorousFlowCapillaryPressureVG
        alpha = 1E-4
        m = 0.9
        pc_max = 1e5
    []
[]

[GlobalParams]
    PorousFlowDictator = dictator
    gravity = '0 0 0'
[]

[Variables]
    [porepressure]
        initial_condition = -2E4
    []
[]

[Kernels]
    [dmdt]
        type = PorousFlowMassTimeDerivative
        variable = porepressure
    []
    [flux]
        type = PorousFlowAdvectiveFlux
        variable = porepressure
    []
[]

[AuxVariables]
    [saturation]
        family = MONOMIAL
        order = CONSTANT
    []
[]

[AuxKernels]
    [saturation]
        type = PorousFlowPropertyAux
        variable = saturation
        property = saturation
    []
[]

[BCs]
    [injection]
        type = PorousFlowPiecewiseLinearSink
        variable = porepressure
        boundary = left
        pt_vals = '-1E9 1E9'
        multipliers = '-1E9 1E9'
        flux_function = 0.046
        PT_shift = 0.35E+5
    []
[]

[Modules]
    [FluidProperties]
        [liquid]
            type = SimpleFluidProperties
            viscosity = 0.06
            density0 = 920
        []
    []
[]

[Materials]
    [porosity_block0]
        type = PorousFlowPorosity
        porosity_zero = 0.7
    []
    [permeability_block0]
        type = PorousFlowPermeabilityConst
        permeability = '3E-10 0 0   0 3E-10 0   0 0 3E-10'
    []
    [saturation_calculator]
        type = PorousFlow1PhaseP
        porepressure = porepressure
        capillary_pressure = pc
    []
    [temperature]
        type = PorousFlowTemperature
        temperature = 293
    []
    [massfrac]
        type = PorousFlowMassFraction
    []
    [liquid]
        type = PorousFlowSingleComponentFluid
        fp = liquid
        phase = 0
    []
    [relperm]
        type = PorousFlowRelativePermeabilityCorey
        n = 3
        s_res = 0.1
        sum_s_res = 0.1
        phase = 0
    []
[]

[Preconditioning]
    [smp]
        type = SMP
        full = true
        petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
        petsc_options_value = ' lu       mumps'
    []
[]

[Executioner]
    type = Transient
    solve_type = 'NEWTON'
    
    end_time = 200
    [TimeStepper]
        type = SolutionTimeAdaptiveDT
        dt = 1E-3
    []

    automatic_scaling = true
[]

[Outputs]
    exodus = true
    print_linear_residuals = false
[]
```

</details>

**Comment by WilkAndy** [comment, marked as answer]
I'm wondering whether this problem really is described by porous-flow physics, and not something else.

But, let's assume PorousFlow is the way to go.  I wonder whether you've actually measured Pc and have data to parameterise it?  If not, i think i'd use 2-phase PorousFlow (more computationally expensive, so slower, sorry) with a (P, S) formulation (`PorousFlow2PhasePS`), and zero capillary pressure (`PorousFlowCapillaryPressureConst` with `pc = 0`).  Then start at saturation=1, and think that saturation=low corresponds to resin.  (Or possibly, start at saturation=0 and think that saturation=high corresponds to resin - whether this is computationally better typically depends on your boundary conditions, but you can easily experiment with both.)   

Alternatively, but this is something i don't think i've tried before so can't guarantee it'll work but it'll be computationally faster than the previous paragraph: use your single-phase approach, but use `pc = 0` (as in previous paragraph) but start at saturation=1, and consider saturation as representing the saturation of the vacuum, so saturation=low means resin.

a

**Reply by s-dunnim** [reply]
Thank you, I'll have a go with both of those suggestions. The literature generally uses Darcy law to model the flow but the effects of capillary pressure are often omitted. Unfortunately I don't (and won't) have any data on Pc, I'm limited really to resin viscosity and density, and fibre porosity and permeability. Hopefully using Pc=0 will help!
